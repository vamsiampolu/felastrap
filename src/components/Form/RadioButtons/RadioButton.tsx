import { useFela } from "react-fela";
import { RadioLabel, type RadioLabelProps } from "./Label";
import { RadioInput, type RadioInputProps } from "./RadioInput";
import { ReactNode } from "react";
import { StyleProps } from "../../common";

type RuleProps = StyleProps<{ reverse: boolean; inline: boolean }>;

const rule = () => ({
  display: "block",
  minHeight: "1.5rem",
  paddingLeft: "1.5rem !important",
  paddingRight: 0,
  marginBottom: "0.125rem",
  textAlign: "left",
});

const reverseRule = ({ reverse }: RuleProps) =>
  reverse
    ? {
        paddingRight: "1.5em",
        paddingLeft: 0,
        textAlign: "right",
      }
    : {
        paddingLeft: "1.5rem !important",
        paddingRight: 0,
        textAlign: "left",
      };

const inlineRule = ({ inline }: RuleProps) =>
  inline
    ? {
        display: "inline-block",
        marginRight: "1rem",
      }
    : {
        display: "block",
      };

export type RadioButtonProps = Omit<
  RadioLabelProps & RadioInputProps,
  "htmlFor" | "children"
> & { label: ReactNode; inline?: boolean };
export const RadioButton = ({
  label,
  className = "",
  reverse = false,
  inline = false,
  value,
  required = false,
  name,
  id,
  onSelect,
  onChange = () => undefined,
  onInput,
  disabled = false,
}: RadioButtonProps) => {
  const { css } = useFela({ reverse, inline });
  const cx = `${css([rule, reverseRule, inlineRule])} ${className}`.trim();
  return (
    <div data-testid="form-control-radio-button" className={cx}>
      <RadioInput
        disabled={disabled}
        reverse={reverse}
        required={required}
        value={value}
        name={name}
        id={id}
        onSelect={onSelect}
        onChange={onChange}
        onInput={onInput}
      />
      <RadioLabel htmlFor={id!} disabled={disabled}>
        {label}
      </RadioLabel>
    </div>
  );
};
