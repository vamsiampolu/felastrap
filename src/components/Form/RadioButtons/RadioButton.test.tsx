import { IRenderer } from "fela";
import { RadioButton } from "./RadioButton";
import {
  Wrapper,
  createRenderer,
  createSnapshot,
  cssReboot,
} from "../../../test-helpers";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

describe("components/Form/RadioButtons/RadioButton", () => {
  let renderer: IRenderer;

  beforeEach(() => {
    renderer = createRenderer();
    cssReboot(renderer);
  });

  afterEach(() => {
    renderer.clear();
  });

  it("should render a radio button control", async () => {
    const component = <RadioButton label="Radio Button" id="foobar" />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    expect(screen.getByTestId("form-control-radio-button")).toBeInTheDocument();
    expect(screen.getByTestId("form-label")).toBeInTheDocument();
    expect(screen.getByTestId("radio-input")).toBeInTheDocument();

    expect(screen.queryByTestId("radio-input")).toHaveAttribute("id", "foobar");
    expect(screen.queryByTestId("form-label")).toHaveAttribute("for", "foobar");

    expect(screen.queryByTestId("form-label")).toHaveTextContent(
      "Radio Button",
    );
    expect(screen.queryByTestId("radio-input")).not.toBeChecked();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a disabled radio button control", async () => {
    const component = <RadioButton disabled label="Radio Button" id="foobar" />;

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);
    expect(screen.queryByTestId("radio-input")).toBeDisabled();

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render a reverse radio button control", async () => {
    const component = <RadioButton reverse label="Radio Button" id="foobar" />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should render an inline radio button control", async () => {
    const component = <RadioButton reverse label="Radio Button" id="foobar" />;
    render(<Wrapper renderer={renderer}>{component}</Wrapper>);

    const snapshot = await createSnapshot({ component, renderer });
    expect(snapshot).toMatchSnapshot();
  });

  it("should invoke the onChange prop when radio button is clicked", async () => {
    const onChange = jest.fn();
    const user = userEvent.setup();

    const component = (
      <RadioButton onChange={onChange} label="Radio Button" id="foobar" />
    );

    render(<Wrapper renderer={renderer}>{component}</Wrapper>);
    const radio = screen.queryByTestId("radio-input");

    await user.click(radio!);
    expect(onChange).toHaveBeenCalled();

    expect(radio).toBeChecked();
  });
});
