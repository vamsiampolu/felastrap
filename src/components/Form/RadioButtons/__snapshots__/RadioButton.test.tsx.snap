// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Form/RadioButtons/RadioButton should render a disabled radio button control 1`] = `
".a {
  display: block;
}
.b {
  min-height: 1.5rem;
}
.c {
  padding-left: 1.5rem !important;
}
.d {
  padding-right: 0;
}
.e {
  margin-bottom: 0.125rem;
}
.f {
  text-align: left;
}
.g {
  float: left !important;
}
.h {
  margin-left: -1.5em;
}
.i {
  flex-shrink: 0;
}
.j {
  width: 1em;
}
.k {
  height: 1em;
}
.l {
  margin-top: 0.25em;
}
.m {
  vertical-align: top;
}
.n {
  appearance: none;
}
.o {
  background-color: #fff;
}
.p {
  background-repeat: no-repeat;
}
.q {
  background-position: center;
}
.r {
  background-size: contain;
}
.s {
  print-color-adjust: exact;
}
.t {
  color-adjust: exact;
}
.v:disabled {
  pointer-events: none;
}
.w:disabled {
  filter: none;
}
.x:disabled {
  opacity: 0.5;
}
.ag:checked {
  background-color: #0d6efd;
}
.ah:checked {
  background-image: url(test-file-stub);
}
.ai:checked {
  border-top-color: #0d6efd;
}
.aj:checked {
  border-right-color: #0d6efd;
}
.ak:checked {
  border-bottom-color: #0d6efd;
}
.al:checked {
  border-left-color: #0d6efd;
}
.am {
  border-top-left-radius: 50%;
}
.an {
  border-top-right-radius: 50%;
}
.ao {
  border-bottom-right-radius: 50%;
}
.ap {
  border-bottom-left-radius: 50%;
}
.aq {
  border-top-width: 1px;
}
.ar {
  border-right-width: 1px;
}
.as {
  border-bottom-width: 1px;
}
.at {
  border-left-width: 1px;
}
.au {
  border-top-style: solid;
}
.av {
  border-right-style: solid;
}
.aw {
  border-bottom-style: solid;
}
.ax {
  border-left-style: solid;
}
.ay {
  border-top-color: #dee2e6;
}
.az {
  border-right-color: #dee2e6;
}
.ba {
  border-bottom-color: #dee2e6;
}
.bb {
  border-left-color: #dee2e6;
}
.bc {
  cursor: default;
}
.bd {
  opacity: 0.5;
}
.be {
  display: inline-block;
}
.bf {
  margin-bottom: 0.5rem;
}
.y:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.z:focus {
  outline-width: 0;
}
.ab:focus {
  border-top-color: #86b7fe;
}
.ac:focus {
  border-right-color: #86b7fe;
}
.ae:focus {
  border-bottom-color: #86b7fe;
}
.af:focus {
  border-left-color: #86b7fe;
}
.u:active {
  filter: brightness(90%);
}


<div data-testid="form-control-radio-button" class="a b c d e f">
  <input
    data-testid="radio-input"
    type="radio"
    class="ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb g h i j k l m n o p q r s t u v w x y z form-check-input"
    id="foobar"
    disabled=""
  /><label
    data-testid="form-label"
    for="foobar"
    class="be bf form-label bc bd form-check-label"
    >Radio Button</label
  >
</div>
;
"
`;

exports[`components/Form/RadioButtons/RadioButton should render a radio button control 1`] = `
".a {
  display: block;
}
.b {
  min-height: 1.5rem;
}
.c {
  padding-left: 1.5rem !important;
}
.d {
  padding-right: 0;
}
.e {
  margin-bottom: 0.125rem;
}
.f {
  text-align: left;
}
.g {
  float: left !important;
}
.h {
  margin-left: -1.5em;
}
.i {
  flex-shrink: 0;
}
.j {
  width: 1em;
}
.k {
  height: 1em;
}
.l {
  margin-top: 0.25em;
}
.m {
  vertical-align: top;
}
.n {
  appearance: none;
}
.o {
  background-color: #fff;
}
.p {
  background-repeat: no-repeat;
}
.q {
  background-position: center;
}
.r {
  background-size: contain;
}
.s {
  print-color-adjust: exact;
}
.t {
  color-adjust: exact;
}
.v:disabled {
  pointer-events: none;
}
.w:disabled {
  filter: none;
}
.x:disabled {
  opacity: 0.5;
}
.ag:checked {
  background-color: #0d6efd;
}
.ah:checked {
  background-image: url(test-file-stub);
}
.ai:checked {
  border-top-color: #0d6efd;
}
.aj:checked {
  border-right-color: #0d6efd;
}
.ak:checked {
  border-bottom-color: #0d6efd;
}
.al:checked {
  border-left-color: #0d6efd;
}
.am {
  border-top-left-radius: 50%;
}
.an {
  border-top-right-radius: 50%;
}
.ao {
  border-bottom-right-radius: 50%;
}
.ap {
  border-bottom-left-radius: 50%;
}
.aq {
  border-top-width: 1px;
}
.ar {
  border-right-width: 1px;
}
.as {
  border-bottom-width: 1px;
}
.at {
  border-left-width: 1px;
}
.au {
  border-top-style: solid;
}
.av {
  border-right-style: solid;
}
.aw {
  border-bottom-style: solid;
}
.ax {
  border-left-style: solid;
}
.ay {
  border-top-color: #dee2e6;
}
.az {
  border-right-color: #dee2e6;
}
.ba {
  border-bottom-color: #dee2e6;
}
.bb {
  border-left-color: #dee2e6;
}
.bc {
  display: inline-block;
}
.bd {
  margin-bottom: 0.5rem;
}
.y:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.z:focus {
  outline-width: 0;
}
.ab:focus {
  border-top-color: #86b7fe;
}
.ac:focus {
  border-right-color: #86b7fe;
}
.ae:focus {
  border-bottom-color: #86b7fe;
}
.af:focus {
  border-left-color: #86b7fe;
}
.u:active {
  filter: brightness(90%);
}


<div data-testid="form-control-radio-button" class="a b c d e f">
  <input
    data-testid="radio-input"
    type="radio"
    class="ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb g h i j k l m n o p q r s t u v w x y z form-check-input"
    id="foobar"
  /><label
    data-testid="form-label"
    for="foobar"
    class="bc bd form-label form-check-label"
    >Radio Button</label
  >
</div>
;
"
`;

exports[`components/Form/RadioButtons/RadioButton should render a reverse radio button control 1`] = `
".a {
  display: block;
}
.b {
  min-height: 1.5rem;
}
.c {
  padding-left: 0;
}
.d {
  padding-right: 1.5em;
}
.e {
  margin-bottom: 0.125rem;
}
.f {
  text-align: right;
}
.g {
  float: left !important;
}
.h {
  margin-left: -1.5em;
}
.i {
  flex-shrink: 0;
}
.j {
  width: 1em;
}
.k {
  height: 1em;
}
.l {
  margin-top: 0.25em;
}
.m {
  vertical-align: top;
}
.n {
  appearance: none;
}
.o {
  background-color: #fff;
}
.p {
  background-repeat: no-repeat;
}
.q {
  background-position: center;
}
.r {
  background-size: contain;
}
.s {
  print-color-adjust: exact;
}
.t {
  color-adjust: exact;
}
.v:disabled {
  pointer-events: none;
}
.w:disabled {
  filter: none;
}
.x:disabled {
  opacity: 0.5;
}
.ag:checked {
  background-color: #0d6efd;
}
.ah:checked {
  background-image: url(test-file-stub);
}
.ai:checked {
  border-top-color: #0d6efd;
}
.aj:checked {
  border-right-color: #0d6efd;
}
.ak:checked {
  border-bottom-color: #0d6efd;
}
.al:checked {
  border-left-color: #0d6efd;
}
.am {
  border-top-left-radius: 50%;
}
.an {
  border-top-right-radius: 50%;
}
.ao {
  border-bottom-right-radius: 50%;
}
.ap {
  border-bottom-left-radius: 50%;
}
.aq {
  border-top-width: 1px;
}
.ar {
  border-right-width: 1px;
}
.as {
  border-bottom-width: 1px;
}
.at {
  border-left-width: 1px;
}
.au {
  border-top-style: solid;
}
.av {
  border-right-style: solid;
}
.aw {
  border-bottom-style: solid;
}
.ax {
  border-left-style: solid;
}
.ay {
  border-top-color: #dee2e6;
}
.az {
  border-right-color: #dee2e6;
}
.ba {
  border-bottom-color: #dee2e6;
}
.bb {
  border-left-color: #dee2e6;
}
.bc {
  display: inline-block;
}
.bd {
  margin-bottom: 0.5rem;
}
.y:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.z:focus {
  outline-width: 0;
}
.ab:focus {
  border-top-color: #86b7fe;
}
.ac:focus {
  border-right-color: #86b7fe;
}
.ae:focus {
  border-bottom-color: #86b7fe;
}
.af:focus {
  border-left-color: #86b7fe;
}
.u:active {
  filter: brightness(90%);
}


<div data-testid="form-control-radio-button" class="a b c d e f">
  <input
    data-testid="radio-input"
    type="radio"
    class="ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb c d f g h i j k l m n o p q r s t u v w x y z form-check-input"
    id="foobar"
  /><label
    data-testid="form-label"
    for="foobar"
    class="bc bd form-label form-check-label"
    >Radio Button</label
  >
</div>
;
"
`;

exports[`components/Form/RadioButtons/RadioButton should render an inline radio button control 1`] = `
".a {
  display: block;
}
.b {
  min-height: 1.5rem;
}
.c {
  padding-left: 0;
}
.d {
  padding-right: 1.5em;
}
.e {
  margin-bottom: 0.125rem;
}
.f {
  text-align: right;
}
.g {
  float: left !important;
}
.h {
  margin-left: -1.5em;
}
.i {
  flex-shrink: 0;
}
.j {
  width: 1em;
}
.k {
  height: 1em;
}
.l {
  margin-top: 0.25em;
}
.m {
  vertical-align: top;
}
.n {
  appearance: none;
}
.o {
  background-color: #fff;
}
.p {
  background-repeat: no-repeat;
}
.q {
  background-position: center;
}
.r {
  background-size: contain;
}
.s {
  print-color-adjust: exact;
}
.t {
  color-adjust: exact;
}
.v:disabled {
  pointer-events: none;
}
.w:disabled {
  filter: none;
}
.x:disabled {
  opacity: 0.5;
}
.ag:checked {
  background-color: #0d6efd;
}
.ah:checked {
  background-image: url(test-file-stub);
}
.ai:checked {
  border-top-color: #0d6efd;
}
.aj:checked {
  border-right-color: #0d6efd;
}
.ak:checked {
  border-bottom-color: #0d6efd;
}
.al:checked {
  border-left-color: #0d6efd;
}
.am {
  border-top-left-radius: 50%;
}
.an {
  border-top-right-radius: 50%;
}
.ao {
  border-bottom-right-radius: 50%;
}
.ap {
  border-bottom-left-radius: 50%;
}
.aq {
  border-top-width: 1px;
}
.ar {
  border-right-width: 1px;
}
.as {
  border-bottom-width: 1px;
}
.at {
  border-left-width: 1px;
}
.au {
  border-top-style: solid;
}
.av {
  border-right-style: solid;
}
.aw {
  border-bottom-style: solid;
}
.ax {
  border-left-style: solid;
}
.ay {
  border-top-color: #dee2e6;
}
.az {
  border-right-color: #dee2e6;
}
.ba {
  border-bottom-color: #dee2e6;
}
.bb {
  border-left-color: #dee2e6;
}
.bc {
  display: inline-block;
}
.bd {
  margin-bottom: 0.5rem;
}
.y:focus {
  box-shadow: 0 0 0 0.25rem rgba(13, 110, 253, 0.25);
}
.z:focus {
  outline-width: 0;
}
.ab:focus {
  border-top-color: #86b7fe;
}
.ac:focus {
  border-right-color: #86b7fe;
}
.ae:focus {
  border-bottom-color: #86b7fe;
}
.af:focus {
  border-left-color: #86b7fe;
}
.u:active {
  filter: brightness(90%);
}


<div data-testid="form-control-radio-button" class="a b c d e f">
  <input
    data-testid="radio-input"
    type="radio"
    class="ab ac ae af ag ah ai aj ak al am an ao ap aq ar as at au av aw ax ay az ba bb c d f g h i j k l m n o p q r s t u v w x y z form-check-input"
    id="foobar"
  /><label
    data-testid="form-label"
    for="foobar"
    class="bc bd form-label form-check-label"
    >Radio Button</label
  >
</div>
;
"
`;
